index.ts

```diff
@@ -152,186 +213,210 @@ module.exports = {};`,
+      if (typeof second === 'string') {
+        let version = second;
+        dep = new Project(name, version, third);
+      } else {
+        dep = new Project(name, second);
+      }
+    } else if ('isDependency' in first) {
+      dep = first;
     } else {
-      throw new TypeError('WTF');
-    }
-
-    if (typeof cb === 'function') {
-      cb(dep);
+      dep = new Project(first);
     }
 
+    this[target][dep.name] = dep;
+    this.dependencyLinks.delete(dep.name);
+    this.linkIsDevDependency.delete(dep.name);
     return dep;
   }
 
   removeDependency(name: string) {
     delete this._dependencies[name];
+    this.dependencyLinks.delete(name);
+    this.linkIsDevDependency.delete(name);
   }
 
   removeDevDependency(name: string) {
     delete this._devDependencies[name];
+    this.dependencyLinks.delete(name);
+    this.linkIsDevDependency.delete(name);
   }
 
-  addDevDependency(name: string | Project, version?: string, cb?: (project: Project) => void) {
-    let dep;
+  addDevDependency(name?: string, version?: string, args?: Omit<ProjectArgs, 'name' | 'version'>): Project;
+  addDevDependency(name?: string, args?: Omit<ProjectArgs, 'name'>): Project;
+  addDevDependency(args?: ProjectArgs): Project;
+  addDevDependency(args?: Project): Project;
+  addDevDependency(
+    first?: string | ProjectArgs | Project,
+    second?: string | Omit<ProjectArgs, 'name'>,
+    third?: Omit<ProjectArgs, 'name' | 'version'>
+  ): Project {
+    return this.addDep(first, second, third, '_devDependencies');
+  }
 
-    if (typeof name === 'string') {
-      dep = this._devDependencies[name] = new (this.constructor as ProjectConstructor)(
-        name,
-        version,
-        undefined,
-        path.join(this.root, this.name, 'node_modules')
-      );
-    } else if (name.isDependency) {
-      dep = this._devDependencies[name.name] = name;
+  linkDependency(
+    name: string,
+    opts:
+      | { baseDir: string; resolveName?: string; requestedRange?: string }
+      | { target: string; requestedRange?: string }
+  ) {
+    this.removeDependency(name);
+    this.removeDevDependency(name);
+    let dir: string;
+    if ('baseDir' in opts) {
+      let pkgJSONPath = resolvePackagePath(opts.resolveName || name, opts.baseDir);
+      if (!pkgJSONPath) {
+        throw new Error(`failed to locate ${opts.resolveName || name} in ${opts.baseDir}`);
+      }
+      dir = path.dirname(pkgJSONPath);
     } else {
-      throw new TypeError('WTF');
-    }
-
-    if (typeof cb === 'function') {
-      cb(dep);
+      dir = opts.target;
     }
+    let requestedRange = opts?.requestedRange ?? fs.readJsonSync(path.join(dir, 'package.json')).version;
+    this.dependencyLinks.set(name, { dir, requestedRange });
+  }
 
-    return dep;
+  linkDevDependency(name: string, opts: { baseDir: string; resolveName?: string } | { target: string }) {
+    this.linkDependency(name, opts);
+    this.linkIsDevDependency.add(name);
   }
 
-  dependencies() {
```

Comment:
no longer exposing `dependency` this way also seems like a high cost API change...